### ISSUE TITLE
"MongoServerSelectionError: Server selection timed out after * ms" not handled

### ISSUE THREAD

**Issue by hasezoey** [nodkz/mongodb-memory-server#431]
## Versions

- NodeJS: 14.15.4
- mongodb-memory-server-core: 7.0.0-beta.14
- system: Linux

package: mongodb-memory-server-core

## What is the Problem?

Error `MongoServerSelectionError: Server selection timed out after * ms` is not handled, which results in an `.start` without end

## Code Example

```ts
import { MongoMemoryReplSet } from "mongodb-memory-server-core";
import * as path from "path";

async function startDB() {
  // note: this path must exists before running this script, otherwise it will error
  const dbPath = path.join(process.cwd(), "db");
  console.log("*dbPath: ", dbPath);
  const replSet = await MongoMemoryReplSet.create({
    replSet: {
      storageEngine: "wiredTiger",
    },
    instanceOpts: [
      {
        port: 9000,
        dbPath,
      },
    ],
  });
  const uri = replSet.getUri();
  console.log("uri: ", uri);

  await replSet.stop();
}

startDB();
```

## Do you know *why* it happenes?

not yet

<details>
<summary>Console Output</summary>

```
*dbPath:  /home/hasezoey/Downloads/typegoose-testing/db
  MongoMS:MongoMemoryReplSet start +0ms
  MongoMS:MongoMemoryReplSet init +1ms
  MongoMS:MongoMemoryReplSet   starting server from instanceOpts (count: 1): { port: 9000, dbPath: '/home/hasezoey/Downloads/typegoose-testing/db' } +0ms
  MongoMS:MongoMemoryReplSet    instance opts: {
  auth: false,
  args: [],
  dbName: '8783248d-463d-4275-aca3-ce7ca370108c',
  ip: '127.0.0.1',
  replSet: 'testset',
  storageEngine: 'wiredTiger',
  port: 9000,
  dbPath: '/home/hasezoey/Downloads/typegoose-testing/db'
} +1ms
  MongoMS:MongoMemoryServer Called MongoMemoryServer.start() method +0ms
  MongoMS:MongoMemoryServer Called MongoMemoryServer._startUpInstance() method +0ms
  MongoMS:MongoMemoryServer getStartOptions +1ms
  MongoMS:MongoMemoryServer Checking if "/home/hasezoey/Downloads/typegoose-testing/db}" (no new tmpDir) already has data +1ms
  MongoMS:MongoMemoryServer Creating new MongoDB instance with options: {"instance":{"dbPath":"/home/hasezoey/Downloads/typegoose-testing/db","ip":"127.0.0.1","port":9000,"storageEngine":"wiredTiger","replSet":"testset","args":[],"auth":false},"binary":{},"spawn":{}} +0ms
  MongoMS:MongoBinary MongoBinary options: {
  "downloadDir": "/home/hasezoey/.cache/mongodb-binaries",
  "platform": "linux",
  "arch": "x64",
  "version": "4.0.20",
  "checkMD5": false
} +0ms
  MongoMS:MongoBinary MongoBinary: Download lock removed +1ms
  MongoMS:MongoBinary MongoBinary: Mongod binary path: "/home/hasezoey/.cache/mongodb-binaries/4.0.20/mongod" +0ms
  MongoMS:MongoInstance Mongo[9000]: prepareCommandArgs +0ms
  MongoMS:MongoInstance Mongo[9000]: prepareCommandArgs: final arugment array:["--port","9000","--dbpath","/home/hasezoey/Downloads/typegoose-testing/db","--storageEngine","wiredTiger","--bind_ip","127.0.0.1","--noauth","--replSet","testset"] +0ms
  MongoMS:MongoInstance Mongo[9000]: Called MongoInstance._launchKiller(parent: 8501, child: 8512): +3ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:28.925+0100 I CONTROL  [initandlisten] MongoDB starting : pid=8512 port=9000 dbpath=/home/hasezoey/Downloads/typegoose-testing/db 64-bit host=Meicoo
  MongoMS:MongoInstance 2021-01-22T10:42:28.925+0100 I CONTROL  [initandlisten] db version v4.0.20
  MongoMS:MongoInstance 2021-01-22T10:42:28.925+0100 I CONTROL  [initandlisten] git version: e2416422da84a0b63cde2397d60b521758b56d1b
  MongoMS:MongoInstance 2021-01-22T10:42:28.925+0100 I CONTROL  [initandlisten] allocator: tcmalloc
  MongoMS:MongoInstance 2021-01-22T10:42:28.925+0100 I CONTROL  [initandlisten] modules: none
  MongoMS:MongoInstance 2021-01-22T10:42:28.925+0100 I CONTROL  [initandlisten] build environment:
  MongoMS:MongoInstance 2021-01-22T10:42:28.925+0100 I CONTROL  [initandlisten]     distarch: x86_64
  MongoMS:MongoInstance 2021-01-22T10:42:28.925+0100 I CONTROL  [initandlisten]     target_arch: x86_64
  MongoMS:MongoInstance  +27ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:28.925+0100 I CONTROL  [initandlisten] options: { net: { bindIp: "127.0.0.1", port: 9000 }, replication: { replSet: "testset" }, security: { authorization: "disabled" }, storage: { dbPath: "/home/hasezoey/Downloads/typegoose-testing/db", engine: "wiredTiger" } }
  MongoMS:MongoInstance 2021-01-22T10:42:28.926+0100 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=15550M,cache_overflow=(file_max=0M),session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),statistics_log=(wait=0),verbose=(recovery_progress),
  MongoMS:MongoInstance  +0ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:29.375+0100 I STORAGE  [initandlisten] WiredTiger message [1611308549:375420][8512:0x7f78c463da40], txn-recover: Main recovery loop: starting at 3/4992 to 4/256
  MongoMS:MongoInstance  +449ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:29.422+0100 I STORAGE  [initandlisten] WiredTiger message [1611308549:422672][8512:0x7f78c463da40], txn-recover: Recovering log 3 through 4
  MongoMS:MongoInstance  +47ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:29.459+0100 I STORAGE  [initandlisten] WiredTiger message [1611308549:459166][8512:0x7f78c463da40], txn-recover: Recovering log 4 through 4
  MongoMS:MongoInstance  +37ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:29.482+0100 I STORAGE  [initandlisten] WiredTiger message [1611308549:482880][8512:0x7f78c463da40], txn-recover: Set global recovery timestamp: 0
  MongoMS:MongoInstance  +23ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.158+0100 I RECOVERY [initandlisten] WiredTiger recoveryTimestamp. Ts: Timestamp(0, 0)
  MongoMS:MongoInstance  +676ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.164+0100 I CONTROL  [initandlisten] 
  MongoMS:MongoInstance 2021-01-22T10:42:30.164+0100 I CONTROL  [initandlisten] ** WARNING: soft rlimits too low. rlimits set to 127881 processes, 1048576 files. Number of processes should be at least 524288 : 0.5 times number of files.
  MongoMS:MongoInstance  +6ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.166+0100 I SHARDING [initandlisten] Marking collection local.system.replset as collection version: <unsharded>
  MongoMS:MongoInstance  +2ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.166+0100 I SHARDING [initandlisten] Marking collection admin.system.roles as collection version: <unsharded>
  MongoMS:MongoInstance  +1ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.167+0100 I SHARDING [initandlisten] Marking collection admin.system.version as collection version: <unsharded>
  MongoMS:MongoInstance  +0ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.168+0100 I SHARDING [initandlisten] Marking collection local.startup_log as collection version: <unsharded>
  MongoMS:MongoInstance  +1ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.168+0100 I FTDC     [initandlisten] Initializing full-time diagnostic data capture with directory '/home/hasezoey/Downloads/typegoose-testing/db/diagnostic.data'
  MongoMS:MongoInstance 2021-01-22T10:42:30.168+0100 I SHARDING [initandlisten] Marking collection local.replset.election as collection version: <unsharded>
  MongoMS:MongoInstance 2021-01-22T10:42:30.168+0100 I SHARDING [initandlisten] Marking collection local.replset.minvalid as collection version: <unsharded>
  MongoMS:MongoInstance 2021-01-22T10:42:30.168+0100 I SHARDING [initandlisten] Marking collection local.replset.oplogTruncateAfterPoint as collection version: <unsharded>
  MongoMS:MongoInstance 2021-01-22T10:42:30.168+0100 I SHARDING [initandlisten] Marking collection local.system.rollback.id as collection version: <unsharded>
  MongoMS:MongoInstance  +0ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.169+0100 I REPL     [initandlisten] Did not find local initialized voted for document at startup.
  MongoMS:MongoInstance  +1ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.170+0100 I REPL     [initandlisten] Rollback ID is 1
  MongoMS:MongoInstance  +1ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.170+0100 I REPL     [initandlisten] Did not find local replica set configuration document at startup;  NoMatchingDocument: Did not find replica set configuration document in local.system.replset
  MongoMS:MongoInstance  +0ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.170+0100 I CONTROL  [LogicalSessionCacheRefresh] Sessions collection is not set up; waiting until next sessions refresh interval: Replication has not yet been configured
  MongoMS:MongoInstance  +0ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.170+0100 I SHARDING [LogicalSessionCacheReap] Marking collection config.system.sessions as collection version: <unsharded>
  MongoMS:MongoInstance 2021-01-22T10:42:30.170+0100 I NETWORK  [initandlisten] waiting for connections on port 9000
  MongoMS:MongoInstance  +0ms
  MongoMS:MongoInstance Mongo[9000]: Instance is ready! +1ms
  MongoMS:MongoMemoryReplSet _initReplSet +1s
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.171+0100 I CONTROL  [LogicalSessionCacheReap] Sessions collection is not set up; waiting until next sessions reap interval: config.system.sessions does not exist
  MongoMS:MongoInstance 2021-01-22T10:42:30.179+0100 I NETWORK  [listener] connection accepted from 127.0.0.1:49730 #1 (1 connection now open)
  MongoMS:MongoInstance  +8ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:30.183+0100 I NETWORK  [conn1] received client metadata from 127.0.0.1:49730 conn1: { driver: { name: "nodejs", version: "3.6.3" }, os: { type: "Linux", name: "linux", architecture: "x64", version: "5.4.0-64-generic" }, platform: "'Node.js v14.15.4, LE (unified)" }
  MongoMS:MongoInstance  +4ms
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:42:31.000+0100 I SHARDING [ftdc] Marking collection local.oplog.rs as collection version: <unsharded>
  MongoMS:MongoInstance  +817ms
(node:8501) UnhandledPromiseRejectionWarning: MongoServerSelectionError: Server selection timed out after 30000 ms
    at Timeout._onTimeout (/home/hasezoey/Downloads/typegoose-testing/node_modules/mongodb/lib/core/sdam/topology.js:438:30)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:8501) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:8501) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
  MongoMS:MongoInstance Mongo[9000]: STDOUT: 2021-01-22T10:43:00.181+0100 I NETWORK  [conn1] end connection 127.0.0.1:49730 (0 connections now open)
  MongoMS:MongoInstance  +29s

^C <-- CTRL+C
```

</details>

**Comment by hasezoey** [comment]
update: from what i can tell - from other issues - is that this error comes if mongodb cant reach any server within 30s (as the errors says, but other issues confirmed)

-> for now i will try to enable more logging around the sometimes failing test for the case it happens again

**Comment by hasezoey** [comment]
this issue may be fixed by e2c55117b1040a70a95f38d42e433ef3fbdd7f04, which adds an extra fail-save to the timeout
-> from what i could tell, is that the timeout didnt get cleared after an promise finished, so added this boolean in the hope that it fixed this

**Comment by hasezoey** [comment]
found the issue, 03184d47e7e3f54d96834711c028a9fb43454441 fixed this, released in `7.0.0-beta.26`
